build_tools/third_party/cuda/CMakeLists.txt

```diff
@@ -4,80 +4,183 @@
+  # of version.txt, which was changed to version.json in recent releases
+  # and thwarts the search.
+  set(IREE_CUDA_LIBDEVICE_PATH "${CUDAToolkit_LIBRARY_ROOT}/nvvm/libdevice/libdevice.10.bc")
+elseif(CUDAToolkit_FOUND AND CUDAToolkit_BIN_DIR)
+  # Back-track from the bin dir as a fallback.
+  set(IREE_CUDA_LIBDEVICE_PATH "${CUDAToolkit_BIN_DIR}/../nvvm/libdevice/libdevice.10.bc")
+elseif(CUDAToolkit_ROOT)
+  # Sometimes the CUDA toolkit doesn't detect... because, you know. Computers
+  # are hard and such. In this case, if the user went to the trouble to
+  # tell us where it is, we have enough information.
+  set(IREE_CUDA_LIBDEVICE_PATH "${CUDAToolkit_ROOT}/nvvm/libdevice/libdevice.10.bc")
+elseif(IREE_CUDA_DOWNLOAD_LIBDEVICE_PATH)
+  message(STATUS "Using downloaded CUDA libdevice")
+  set(IREE_CUDA_LIBDEVICE_PATH "${IREE_CUDA_DOWNLOAD_LIBDEVICE_PATH}")
+else()
+  message(FATAL_ERROR "Building with IREE_TARGET_BACKEND_CUDA requires either a CUDA SDK (consult CMake docs for your version: https://cmake.org/cmake/help/latest/module/FindCUDAToolkit.html) or an explicit path to libdevice (set with -DIREE_CUDA_LIBDEVICE_PATH=/path/to/libdevice.10.bc)")
 endif()
 
-# Then use the download script to fetch and flatten each component we want
-# into the tmp dir.
-# This will produce a unified directory tree under:
-#   flat/$OS-$ARCH
-set(SRC_DIR "${_DOWNLOAD_DIR}/${_OS}-${_ARCH}")
-foreach(COMPONENT ${_COMPONENTS_FO_FETCH})
-  message(STATUS "Downloading component ${COMPONENT}")
-  execute_process(COMMAND ${Python3_EXECUTABLE} "${_DOWNLOAD_SCRIPT_PATH}"
-    --label "${_VERSION}"
-    --product "${_PRODUCT}"
-    --os "${_OS}"
-    --arch "${_ARCH}"
-    --component "${COMPONENT}"
-    --output "${_DOWNLOAD_DIR}")
-endforeach()
-
-if(NOT EXISTS "${SRC_DIR}")
-  message(FATAL_ERROR "Download did not produce expected source dir: ${SRC_DIR}")
-  return()
+if(EXISTS "${IREE_CUDA_LIBDEVICE_PATH}")
+  message(STATUS "Using CUDA libdevice: ${IREE_CUDA_LIBDEVICE_PATH}")
+else()
+  message(SEND_ERROR "Cannot find CUDA libdevice file (${IREE_CUDA_LIBDEVICE_PATH}). Either configure your CUDA SDK such that it can be found or specify explicitly via -DIREE_CUDA_LIBDEVICE_PATH=/path/to/libdevice.10.bc")
 endif()
 
-foreach(REL_PATH ${_RETAIN_PATHS})
-  set(SRC_FILE "${SRC_DIR}/${REL_PATH}")
-  set(TARGET_FILE "${_TARGET_DIR}/${REL_PATH}")
-  message(STATUS "Copy ${SRC_FILE} -> ${TARGET_FILE}")
-  # file(COPY) expects a destination directory.
-  get_filename_component(TARGET_DIR "${TARGET_FILE}" DIRECTORY)
-  file(COPY ${SRC_FILE} DESTINATION ${TARGET_DIR})
-endforeach()
-
-# Delete tmp directory.
-file(REMOVE_RECURSE ${_DOWNLOAD_DIR})
-
-# Set vars for downloaded cuda deps
-set(IREE_CUDA_DOWNLOAD_LIBDEVICE_PATH
-    "${_TARGET_DIR}/nvvm/libdevice/libdevice.10.bc" PARENT_SCOPE)
-set(IREE_CUDA_DOWNLOAD_INCLUDE_PATH "${_TARGET_DIR}/include" PARENT_SCOPE)
+# Locate runtime components.
+if(CUDAToolkit_FOUND)
```

Comment:
Shouldn't we gate this on `IREE_HAL_DRIVER_CUDA`? Similarly, the above should be gated on `IREE_TARGET_BACKEND_CUDA`. I think putting things in functions would make those conditionals a bit more readable

Reply:
Same comment about top-level (all of this came from the project level CMakeLists.txt).

I deliberately simplified this in order to reduce one degree of freedom, but the point is debatable. I think it is ok for us to just say that these are our minimum requirements for "having CUDA" available. It technically could be restricted a bit more but in my experience, a lot of optionality on this dep specifically is the enemy and I think it is fine to just unconditionally try to configure what we find.

Reply:
Ok. I'm fine with "if you're using Cuda you need to go full Cuda", as long as if you're not using Cuda you don't have to deal with it.